File: src/main/java/minic/uilocal/visual/MiniCAssemblyTextLine.java
package minic.uilocal;

import minic.uiapi.UiSourceSpanDto;

import java.util.Objects;

/**
 * Assembly 文本视图中的一行。
 *
 * @param lineNumber 行号
 * @param text 汇编文本
 * @param section section 元信息
 * @param label label 元信息
 * @param kind 行类型
 * @param range 对应源码范围；没有直接对应时为 {@code null}
 * @param active 是否当前行
 */
public record MiniCAssemblyTextLine(
        int lineNumber,
        String text,
        String section,
        String label,
        String kind,
        UiSourceSpanDto range,
        boolean active
) {
    public MiniCAssemblyTextLine {
        Objects.requireNonNull(text, "text");
        Objects.requireNonNull(section, "section");
        Objects.requireNonNull(label, "label");
        Objects.requireNonNull(kind, "kind");
        if (lineNumber < 1) {
            throw new IllegalArgumentException("lineNumber must be 1-based");
        }
    }
}

File: uiweb/src/visual/MiniCAssemblyTextLine.ts
import type { JavaMirrorFile, UiSourceSpanDto } from "../translation/javaMirror";

export const miniCAssemblyTextLineMirror = {
  javaPath: "src/main/java/minic/uilocal/visual/MiniCAssemblyTextLine.java",
  webPath: "uiweb/src/visual/MiniCAssemblyTextLine.ts",
  packageName: "minic.uilocal.visual",
  exportName: "MiniCAssemblyTextLine",
  kind: "record",
  imports: ["minic.uiapi.UiSourceSpanDto", "java.util.Objects"],
  fields: [],
  methods: [{ name: "MiniCAssemblyTextLine", signature: "MiniCAssemblyTextLine(int lineNumber, String text, String section, String label, String kind, UiSourceSpanDto range, boolean active)" }],
} as const satisfies JavaMirrorFile;

export class MiniCAssemblyTextLine {
  static readonly mirror = miniCAssemblyTextLineMirror;

  constructor(
    readonly lineNumber: number,
    readonly text: string,
    readonly section: string,
    readonly label: string,
    readonly kind: string,
    readonly range: UiSourceSpanDto | null,
    readonly active: boolean,
  ) {
    if (lineNumber < 1) {
      throw new Error("lineNumber must be 1-based");
    }
  }
}

export default MiniCAssemblyTextLine;
